exercises/flowcache/mycontroller.py

```diff
@@ -138,27 +162,76 @@ def addFlowRule( ingress_sw, src_ip_addr, dst_ip_addr, protocol, port, new_dscp,
-            "hdr.ipv4.src_addr": src_ip_addr,
-            "hdr.ipv4.dst_addr": dst_ip_addr
+            "hdr.ipv4.srcAddr": src_ip_addr,
+            "hdr.ipv4.dstAddr": dst_ip_addr
         },
-        action_name="ingressImpl.cached_action",
+        action_name="MyIngress.cached_action",
         action_params={
             "port":           port,
             "decrement_ttl":  x,
             "new_dscp":       new_dscp,
+            "dst_eth_addr":   dst_eth_addr
         },
-        idle_timeout_ns = 3
+        # TODO: Add idle timeout
         )
     ingress_sw.WriteTableEntry(table_entry)
-    print("Installed ingress tunnel rule on %s" % ingress_sw.name)
+    print("Installed ingress rule on %s" % ingress_sw.name)
 
 def sendPacketOut(sw ,payload, metadatas):
-        sw.PacketOut(payload, metadatas)
+    # TODO: Implement the function logic to send a packet-out message
+
+async def readTableRules(p4info_helper, sw):
+    """
+    Reads the table entries from all tables on the switch.
+
+    :param p4info_helper: the P4Info helper
+    :param sw: the switch connection
+    """
+    print('\n----- Reading tables rules for %s -----' % sw.name)
+    for response in sw.ReadTableEntries():
+        for entity in response.entities:
+            entry = entity.table_entry
+            # TODO For extra credit, you can use the p4info_helper to translate
+            #      the IDs in the entry to names
+            print(entry)
+            print('-----')
+
+async def printCounter(p4info_helper, sw, counter_name, index):
+    """
+    Reads the specified counter at the given index from the switch. In our
+    program, the index is derived from the first 6 bits of the IP destination address.
+    If the index is 0, it will return all values from the counter.
+
+    :param p4info_helper: the P4Info helper
+    :param sw:  the switch connection
+    :param counter_name: the name of the counter from the P4 program
+    :param index: the counter index (in our case, first 6 bits of the IP)
+    """
+    try:
+        for response in sw.ReadCounters(p4info_helper.get_counters_id(counter_name), index):
+            for entity in response.entities:
+                counter = entity.counter_entry
+                print("%s %s %d: %d packets (%d bytes)" % (
+                    sw.name, counter_name, index,
+                    counter.data.packet_count, counter.data.byte_count
+                ))
+    except grpc.RpcError as e:
+           print(f"[gRPC Error in printCounter for {sw.name}]")
+           printGrpcError(e)
+
+           if e.code() == grpc.StatusCode.UNKNOWN:
+            print(f"Unknown gRPC error from {sw.name}. Retrying...")
+           await asyncio.sleep(2)
 
-async def process_packet(message):
+    except Exception as e:
+           print(f"[Unexpected Error in printCounter for {sw.name}]: {e}")
+           traceback.print_exc()
+           await asyncio.sleep(2)
+
+async def processPacket(message):
```

Comment:
The line shortly after this comment:
```
        print("Received %d PacketIn messages" % (len(payload)))
```
is confusing.  When I see "Received 52 PacketIn messages" printed by mycontroller.py, it suggests that it received 52 separate PacketIn messages, when in fact it has received only 1, with a packet length of 52 bytes.